### PR TITLE
Next.js: Fix webpack fsCache not working

### LINKED ISSUE #29621: [Bug]: Webpack fsCache does not work with @storybook/nextjs framework
### Describe the bug

The [Storybook Next.js framework](https://storybook.js.org/docs/get-started/frameworks/nextjs) uses webpack internally to build Storybooks for Next.js projects.

The [Storybook webpack builder](https://storybook.js.org/docs/builders/webpack) supports an `fsCache` option to enable disk-based caching, which greatly improves incremental build performance in webpack-based Storybook projects.

When the `fsCache` option is enabled in a Storybook project that uses the Next.js framework, however, the disk-based cache does not work, and the Storybook build outputs a large number of `webpack.cache.packFileCacheStrategy` warnings. As a result, Next.js projects cannot benefit from the significant speed boost afforded by the webpack disk cache.

#### Storybook config

```
import { StorybookConfig } from "@storybook/nextjs"

export default {
  stories: ["../example/**/*.stories.(ts|tsx)"],
  framework: {
    name: "@storybook/nextjs",
    options: {
      builder: {
        fsCache: true,
      },
    },
  },
  addons: [],
} satisfies StorybookConfig
```

### Reproduction link

https://github.com/sentience/storybook-nextjs-webpack-cache-issue-demo

### Reproduction steps

1. `git clone https://github.com/sentience/storybook-nextjs-webpack-cache-issue-demo.git`
2. `cd storybook-nextjs-webpack-cache-issue-demo`
3. `pnpm i`
4. `pnpm storybook dev`

### System

```bash
Storybook Environment Info:

  System:
    OS: macOS 15.1
    CPU: (8) arm64 Apple M1
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.19.1 - ~/Developer/storybook-nextjs-webpack-cache-issue-demo/.devbox/nix/profile/default/bin/node
    Yarn: 1.22.22 - ~/Developer/storybook-nextjs-webpack-cache-issue-demo/.devbox/virtenv/nodejs/corepack-bin/yarn
    npm: 10.2.4 - ~/Developer/storybook-nextjs-webpack-cache-issue-demo/.devbox/nix/profile/default/bin/npm
    pnpm: 9.12.3 - ~/Developer/storybook-nextjs-webpack-cache-issue-demo/.devbox/virtenv/nodejs/corepack-bin/pnpm <----- active
  Browsers:
    Chrome: 130.0.6723.119
    Safari: 18.1
  npmPackages:
    @storybook/global: ^5.0.0 => 5.0.0 
    @storybook/nextjs: ^8.4.3 => 8.4.3 
    @storybook/types: ^8.4.3 => 8.4.3 
    storybook: ^8.4.3 => 8.4.3
```

### Additional context

This problem seems to have existed since the introduction of @storybook/nextjs in Storybook 8.0.

### PR DESCRIPTION
**Pull request by sentience**
Closes #29621

## What I did

Next.js overrides certain internal webpack packages (particularly `webpack-sources`), which are involved with filesystem caching, with its own versions from next/dist/compiled.

For filesystem caching to work, Next.js must be allowed to perform these overrides before webpack is first initialized by @storybook/builder-webpack5. If it is not, the objects to be serialized to disk in the caching process will be instantiated using the original (non-Next.js) modules, but the cache serializers will be created using the Next.js modules. This mismatch between the objects to be cached and the serializers that write the filesystem cache prevents the cache from being written; instead, webpack outputs a warning message to the console for every object that it tries and fails to find a matching serializer for.

This fix works by invoking Next.js to configure webpack in the `core` hook of @storybook/nextjs/preset, immediately before loading @storybook/builder-webpack5. We don't actually use this configuration that Next.js creates; the actual configuration that will be used in the build is still generated in `webpackFinal` as before.

`fsCache` ([docs](https://storybook.js.org/docs/builders/webpack#configure)) has a large impact on Storybook build performance. Even in a minimal project with a single story, enabling it reduces hot-cache build time by 66%! This is therefore a very valuable option to be able to enable.

### Testing

@storybook/nextjs does not appear to have a test suite, so I don't see an easy way to test this feature. I've verified that it resolves the issue in my repro project for #29621.

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Clone https://github.com/sentience/storybook-nextjs-webpack-cache-issue-demo
2. Run Storybook (`pnpm i`, `pnpm storybook build`) and note the console warnings at the end of the build that indicate the filesystem cache is not working.
3. Upgrade @storybook/nextjs to a canary release with this change.
4. Run Storybook again and note the console warnings are gone, and the build runs in 33% of the time!

### Documentation

- [ ] ~Add or update documentation reflecting your changes~
- [ ] ~If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)~

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-29654-sha-83a1f24f`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-29654-sha-83a1f24f sandbox` or in an existing project with `npx storybook@0.0.0-pr-29654-sha-83a1f24f upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-29654-sha-83a1f24f`](https://npmjs.com/package/storybook/v/0.0.0-pr-29654-sha-83a1f24f) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [sentience/storybook](https://github.com/sentience/storybook) |
| **Branch** | [`fix-nextjs-webpack-fscache`](https://github.com/sentience/storybook/tree/fix-nextjs-webpack-fscache) |
| **Commit** | [`83a1f24f`](https://github.com/sentience/storybook/commit/83a1f24f02c50a0b5ed4a03c6b248100183a3b01) |
| **Datetime** | Tue Dec  3 06:19:41 UTC 2024 (`1733206781`) |
| **Workflow run** | [12134081252](https://github.com/storybookjs/storybook/actions/runs/12134081252) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=29654`_
</details>

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.7 MB | 77.7 MB | 0 B | 0.65 | 0% |
| initSize |  133 MB | 133 MB | -625 B | 0.66 | 0% |
| diffSize |  55.3 MB | 55.3 MB | -625 B | 0.66 | 0% |
| buildSize |  6.87 MB | 6.87 MB | 41 B | -0.03 | 0% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | 0 B | -0.42 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.86 MB | 1.86 MB | 0 B | **1.52** | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.57 MB | 3.57 MB | 0 B | -0.05 | 0% |
| buildPreviewSize |  3.3 MB | 3.3 MB | 41 B | 0.29 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.5s | 6.7s | -772ms | -0.95 | -11.5% |
| generateTime |  20.3s | 18.2s | -2s -171ms | -0.86 | -11.9% |
| initTime |  13.1s | 12s | -1s -126ms | -1.01 | -9.4% |
| buildTime |  9.5s | 7.9s | -1s -596ms | **-1.69** | 🔰-20.1% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5s | 4.6s | -312ms | -0.73 | -6.7% |
| devManagerResponsive |  3.7s | 3.5s | -222ms | -0.69 | -6.3% |
| devManagerHeaderVisible |  549ms | 539ms | -10ms | -0.5 | -1.9% |
| devManagerIndexVisible |  626ms | 570ms | -56ms | -0.89 | -9.8% |
| devStoryVisibleUncached |  1.8s | 1.7s | -112ms | -0.46 | -6.5% |
| devStoryVisible |  624ms | 625ms | 1ms | -0.21 | 0.2% |
| devAutodocsVisible |  507ms | 499ms | -8ms | -0.16 | -1.6% |
| devMDXVisible |  483ms | 465ms | -18ms | -0.61 | -3.9% |
| buildManagerHeaderVisible |  585ms | 515ms | -70ms | -0.74 | -13.6% |
| buildManagerIndexVisible |  692ms | 612ms | -80ms | -0.7 | -13.1% |
| buildStoryVisible |  534ms | 478ms | -56ms | -0.65 | -11.7% |
| buildAutodocsVisible |  410ms | 415ms | 5ms | -0.5 | 1.2% |
| buildMDXVisible |  415ms | 423ms | 8ms | -0.37 | 1.9% |

## Greptile Summary

This PR fixes webpack's filesystem caching in @storybook/nextjs by ensuring Next.js can properly override webpack internals before Storybook's webpack builder initialization, significantly improving build performance.

- Modified `code/frameworks/nextjs/src/preset.ts` to load Next.js configuration in the `core` hook before webpack initialization
- Added early webpack configuration setup to allow Next.js to override internal packages like `webpack-sources`
- Enables proper filesystem caching functionality, reducing build times by up to 66%
- Maintains existing webpack configuration generation in `webpackFinal` hook for actual builds

### CODE REVIEW

File: code/frameworks/nextjs/src/preset.ts [hunk 1]
```diff
@@ -36,6 +36,19 @@ export const addons: PresetProperty<'addons'> = [
...
+  // (see next/dist/server/config-utils.js) before @storybook/builder-webpack5
+  // starts to use it. Without this, webpack's file system cache (fsCache: true)
+  // does not work.
+  const { nextConfigPath } = await options.presets.apply<FrameworkOptions>('frameworkOptions');
+  await configureConfig({
+    // Pass in a dummy webpack config object for now, since we don't want to
+    // modify the real one yet. We pass in the real one in webpackFinal.
+    baseConfig: {},
+    nextConfigPath,
+  });
```
[the remark this answers is not in the document]
Comment (sentience) [reply code/frameworks/nextjs/src/preset.ts:50]: I think the duplication is clearer.

**Comment by sentience** [comment]
@valentinpalkovic May I request a canary release of this? Sorry if you're not the right person to mention; the PR template says to mention @storybook/core, but that doesn't seem to be recognised by GitHub.

**Comment by valentinpalkovic** [comment]
Here we go: https://github.com/storybookjs/storybook/actions/runs/12026402956/job/33525318348

When the job is done, you can see the canary version :)

**Comment by sentience** [comment]
@valentinpalkovic Thanks to that canary build I found some more places where webpack was being referenced before Next.js replaced it with its own internal instance, causing disk cache serializer mismatch errors. I believe I've now fixed all of these.

Could you roll me another canary build, please? 🙏🏻

**Comment by valentinpalkovic** [comment]
Sure :)
https://github.com/storybookjs/storybook/actions/runs/12134081252

**Comment by sentience** [comment]
I've now tested this in several projects, and I think this is in good enough shape to release. Setting `fsCache: true` ([docs](https://storybook.js.org/docs/builders/webpack#configure)) consistently cuts incremental build times by ~66%.

I am, however, aware of one caveat to this fix: projects that use `react-docgen-typescript` to parse React components to generate high-quality props metadata ([docs](https://storybook.js.org/docs/api/main-config/main-config-typescript#reactdocgen)).

If you set `reactDocgen` in your Storybook config like this:

```
/* .storybook/main.ts */
export default {
  ...
  typescript: {
    reactDocgen: "react-docgen-typescript",
  }
};
```

…and set `fsCache: true`, you will still get cache warning messages like this:

> &lt;w> [webpack.cache.PackFileCacheStrategy] Skipped not serializable cache item 'Compilation/modules|/Users/kyank/Developer/unified-home/node_modules/.pnpm/@storybook+nextjs@8.4.6_@swc+core@1.7.21_@swc+helpers@0.5.12__@types+webpack@5.28.5_@swc+core_2ylubrpntvrhqftmmdfysdmhxe/node_modules/@storybook/nextjs/dist/swc/next-swc-loader-patch.js??ruleSet[1].rules[18].use!/Users/kyank/Developer/unified-home/apps/unified-home/src/components/HomeTitleBlock/HomeTitleBlock.tsx': No serializer registered for DocGenDependency
> &lt;w> while serializing webpack/lib/cache/PackFileCacheStrategy.PackContentItems -> webpack/lib/NormalModule -> Array { 19 items } -> DocGenDependency

These messages complain that `DocGenDependency` objects cannot be cached, because they do not have a cache serializer registered. I have traced the root cause of this to the @storybook/react-docgen-typescript-plugin package. It includes [a pair of `import`s](https://github.com/storybookjs/react-docgen-typescript-plugin/blob/026114da4534c6c42b66a9cd169292b620f34d2f/src/dependency.ts#L4-L10) that reach deeply into `webpack`:

```ts
import makeSerializable from "webpack/lib/util/makeSerializable.js";

import NullDependency from "webpack/lib/dependencies/NullDependency.js";
```

[Next.js overrides a bunch of webpack import paths](https://github.com/vercel/next.js/blob/c9aae2f3e9a60c4a2bab6ad41e76897127f51a5a/packages/next/src/server/config-utils.ts#L11-L143) to redirect consumers to its internal copy of webpack, but these paths are not among them.

As a result, @storybook/react-docgen-typescript-plugin registers its cache serializer for `DocGenDependency` with the "original" webpack, instead of Next.js's internal webpack, and this causes the "not serializable cache item" warnings.

This should be possible to fix by updating @storybook/react-docgen-typescript-plugin to use the officially exposed API for registering cache serializers, rather than its current approach that depends on internals that are not exposed by Next.js's webpack instance. Looking at mini-css-extract-plugin, for example it [uses the `webpack.util.serialization.register` API](https://github.com/webpack-contrib/mini-css-extract-plugin/blob/3effaa0319bad5cc1bf0ae760553bf7abcbc35a4/src/index.js#L358) that is [officially documented](https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/src/index.js#L358). @storybook/react-docgen-typescript-plugin looks like it would require some refactoring to get the `webpack` object into scope where needed to use this API, but this should be possible.

In the meantime, however, this PR is no _worse_ than the current behaviour (where _all_ objects print a "not serializable cache item" warning), and even in projects where react-docgen-typescript is used, the partial fix of this PR provides a large boost to incremental build performance.

I therefore propose that this be merged as-is, and I'll open a separate issue to track the warnings in react-docgen-typescript projects.

**Comment by sentience** [comment]
Not sure why `yarn install` failed in CI. Seems to work for me locally.

**Comment by sentience** [comment]
Anything else I can do to help move this along?

**Comment by valentinpalkovic** [comment]
Hi @sentience 

Thank you for your contribution. I was on vacation but I am back again. I'll take a close look this week.

**Pull request merged**